gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+    }
+    String runId = !internalRow.isNullAt(6) ? internalRow.getString(6) : null;
+    boolean dryRun = !internalRow.isNullAt(7) && internalRow.getBoolean(7);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    runId = updateRunId(gcOutputTableIdentifier, runId, identifiedResultsRepo);
+
+    Map<String, List<Long>> expiredSnapshotsPerContentId =
+        getExpiredSnapshotsPerContentId(runId, identifiedResultsRepo);
+
+    Map<String, List<String>> metadataLocationsPerContentId =
+        getMetadataLocationsPerContentId(runId, identifiedResultsRepo);
+
+    List<InternalRow> outputRows = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(nessieClientConfig)) {
+      createGCBranch(expiryBranchName, api);
+      // get Nessie catalog
+      Catalog nessieCatalog = GCUtil.loadNessieCatalog(spark(), gcCatalogName, expiryBranchName);
+      String finalRunId = runId;
+      expiredSnapshotsPerContentId
+          .entrySet()
+          .forEach(
+              entry -> {
+                // Expiry Algorithm:
+                // A. For each content id, fetch ANY one of the content.
+                // B. Use the newly created expiry branch. On this branch,
+                // C. Do a single PUT operation with a fetched content from step-A and
+                //    a dummy ContentKey.
+                // D. Collect the list of all the expired snapshot for this content id.
+                // E. Fetch this table using dummy ContentKey on expiry branch. For this table,
+                // F. call expire_snapshot action from Iceberg by passing the list of expired
+                //    snapshots.
+                // G. Once all the content ids are handled, drop this branch.
+                //    (Else it may impact the next GC operations as some contents are still referred
+                // in
+                //    this branch)
+                String contentId = entry.getKey();
+                addOneContentToExpireProcedureBranch(
+                    expiryBranchName, identifiedResultsRepo, api, finalRunId, contentId);
+                Table table =
+                    nessieCatalog.loadTable(
+                        TableIdentifier.parse(EXPIRE_NAMESPACE + "." + contentId));
+
+                // TODO: Remove after https://github.com/apache/iceberg/pull/4509
+                updateMetadataLocation((BaseTable) table);
+
+                Set<Snapshot> snapshotSet =
+                    new TreeSet<>((o1, o2) -> (int) (o1.timestampMillis() - o2.timestampMillis()));
+                Set<String> processedMetadataFiles = new HashSet<>();
+                List<String> metadataLocations =
+                    new ArrayList<>(metadataLocationsPerContentId.get(contentId));
+
+                metadataLocations.sort(Collections.reverseOrder());
```

Comment:
Why does this need to be sorted?

Reply:
there are lot of validations in table metadata, it expects new snapshot should be younger than previous. So, this validation

Reply:
I don't understand why the order of metadataLocations matters.

Reply:
oh.. This one is an optimization from me.

for content1 from all the branches, expired contents are...
```
idx - localtionX.
idy - locationY.
idx - localtionX.
.
.
idz - locationZ.

```
So, I want to start from the latest version of the metadata files. (hence, the sort)
because it contains .previousFiles(). I can skip those contents metadata file reading if those files are already processed.  The recent file have the most previousFiles().


P.S: 
I will add more comments. I didn't expected this code to be reviewed today. this PR was only to get the overall idea and approach.